Cabal/Distribution/Types/VersionRange.hs

```diff
@@ -469,6 +479,31 @@ instance Parsec VersionRange where
                 (orLaterVersion u) (earlierVersion (majorUpperBound u))
             embed vr = embedVersionRange vr
 
+        -- version set notation (e.g. "== { 0.0.1.0, 0.0.2.0, 0.1.0.0 }")
+        verSet' op vs = do
+            csv <- askCabalSpecVersion
+            if csv >= CabalSpecV3_0
+            then pure $ foldr1 unionVersionRanges (map op vs)
+            else fail $ unwords
+                [ "version set syntax used."
+                , "To use this syntax the package needs to specify at least 'cabal-version: 3.0'."
+                , "Alternatively, if broader compatibility is important then use"
+                , "a series of single version constraints joined with the || operator:"
+                , prettyShow (foldr1 unionVersionRanges (map op vs))
+                ]
+
+        verSet :: CabalParsing m => m [Version]
+        verSet = do
+            _ <- P.char '{'
+            vs <- P.sepBy1 (P.spaces *> verPlain) (P.try (P.spaces *> P.char ','))
+            P.spaces
+            _ <- P.char '}'
+            pure vs
+
+        -- plain version without tags or wildcards
+        verPlain :: CabalParsing m => m Version
+        verPlain = mkVersion <$> P.sepBy1 P.integral (P.char '.')
+
         -- either wildcard or normal version
         verOrWild :: CabalParsing m => m (Bool, Version)
         verOrWild = do
```

Comment:
@hvr if you look at this, this doesn't differentiate between `01.02.03` and `1.2.3` given different `cabal-version`.

I'll change that,  hopefully nothing slipped on Hackage

http://hackage.haskell.org/package/Cabal-2.4.1.0/docs/src/Distribution.Types.Version.html#line-99

Reply:
See also #5138 where we took some measures to prevent denorm versions hitting hackage